### ISSUE TITLE
Focus issue when pressing F10 during an assessment

### ISSUE THREAD

**Issue by chrisswan-rm** [SafeExamBrowser/seb-win-refactoring#408]
We have discovered an issue in the latest version of Safe Exam Browser for Windows (3.3.2), though it also appears to affect 3.2.x, 3.1.x and probably older versions as well.

The assessment we are running involves playing a number of audio clips automatically in a sequence / "playlist" in order to present stimuli for a Listening or Speaking test to a student. The assessment is entirely web-based and uses HTML audio support to play the clips. When running it in SEB, if the student presses F10, the next audio clip in the sequence does not play (or occasionally the currently-playing clip will pause). Usually the sequence resumes if the student presses a key or clicks anywhere within the assessment.

- This happens regardless of whether the F10 key is blocked in the SEB config or not.
- This does not seem to be specific to the audio player we are using, as we have examples of the same behaviour when using JPlayer (https://jplayer.org/), OzPlayer (https://www.accessibilityoz.com/ozplayer/) and when directly using AudioContext. Therefore it appears to be a general issue affecting all audio that is played programmatically.
- We have also seen instances without audio where a page similarly seems to "freeze" when pressing F10, and execution of Javascript only resumes once the student presses a key or clicks on the page.

My guess is that the F10 key is somehow causing SEB to take focus away from the assessment, which results in certain Javascript not executing and audio not playing until the student interacts with the page again. Similar behaviour can be seen in Chrome if you go to a different tab, and this is expected (Chrome doesn't play audio or execute certain Javascript on inactive tabs). The difference in SEB though is that it doesn't look like focus has been taken away, it seems to have been taken away "invisibly".

**Steps to Reproduce**
I have attached two config files where this can be reproduced, to prove that the issue occurs in two totally separate web applications from different vendors.
[seb-f10-issue.zip](https://github.com/SafeExamBrowser/seb-win-refactoring/files/8678701/seb-f10-issue.zip)

For each one, simply run the assessment directly from the config file. The exit code is "1234". To reproduce the issue:

1. Click anywhere in the assessment.
2. Audio clips will play automatically. As each audio clip plays, press F10 once.
3. This will usually either cause a clip to stop playing, or will let the current clip finish and then not allow the next one to play (in the example where the audio controls are visible you will notice the "Play" icon change to a "Pause" icon so it looks like it's going to play, but the clip doesn't actually start).
4. Press a key or click in the assessment to resume. Sometimes it may be necessary to manually press play on the affected clip in the example where the audio controls are visible.

**Expected Behavior**
F10 should do nothing (especially if it's blocked in the SEB config file). 

**Version Information**
 - OS: Windows 10 Enterprise
 - SEB-Version: 3.3.2 (but older versions also affected).

**Comment by chrisswan-rm** [comment]
Thanks Damian, good to hear that you can reproduce it. Did you mean that you can reproduce the same problem (next audio clip not playing) in Firefox and Edge? I tried in Edge, and if I press F10, the "..." menu button in the top right becomes focused, but this doesn't have any adverse effect on the website - the audio clips still continue to play through as expected.

Also, just to clarify, we are not **expecting** the student to press F10 - but just in case they do, it should have no impact on the assessment at all. This issue was picked up in testing of an assessment, and due to this issue it was found that a student can effectively pause the assessment if they press F10, which is not desirable.

F10 appears to have no effect in SEB (even if the menu bar is enabled, F10 doesn't seem to select anything on it), so is it possible to just block it entirely? Or failing that, block it if it's been disabled in the SEB config file.

I suspect that the warning "The AudioContext was not allowed to start" in the log is relevant - did this come up as soon as the site loaded, or a bit later on after pressing F10? If it happened after pressing F10, this looks like more evidence that the issue is that focus has somehow gone away from the page, and the user needs to interact with it again in order for audio to continue playing as expected.

**Comment by s-spann** [comment]
Hi Damian,

Following up on this – we had already disabled this feature in the configuration file however the issue Chris highlighted in the ticket came up despite this being disabled.

This is quite a concern for the business as this allows manipulation of the exam. Do you have any other thoughts?

Kind regards,

Sophie

Sophie Spann
Product Platform Manager
ELT Division | Oxford University Press
Great Clarendon Street | Oxford | OX2 6DP

**Comment by danschlet** [comment]
Are you maybe using some non-standard drivers which could cause this? If we cannot reproduce it on our Windows installations, then there probably is some difference in the system configuration. Or maybe the keyboard layout you're using has different key mappings for F10 than the one we're using.

**Comment by s-spann** [comment]
Hi Daniel, 

I believe this was tested and replicated with two different Windows 10 Enterprise machines with no specific additional driver or keyboard layout. 

The device tested on within our organisation is a Lenovo Thinkpad with an 8th gen core i5 processor. No specific drivers that we are aware of. The "special function" for the F10 key on these keyboards is to enable or disable bluetooth which does not seem to correlate with this issue. 

@chrisswan-rm you may be best placed to advise whether the above applies to your machine also.

**Comment by danschlet** [comment]
Thanks for clarification, @s-spann. I assume you're in the US? Maybe we should use the US keyboard layout and locale trying to replicate this issue, just to dismiss the keyboard layout theory.

**Comment by chrisswan-rm** [comment]
Hi Daniel, we are all testing from the UK with UK keyboards. On my keyboard the special function of F10 is to switch to the view in Windows that shows all of your open applications. But as "Function Lock" was disabled, this feature was not in use at the time I was testing, and F10 should have truly been acting as F10.

I will record a video with exact steps to reproduce it and send you a link.

**Comment by danschlet** [comment]
Sorry for the wrong assumption (actually time zones would indicate you're in Europe). A video would be very helpful in this case, thanks!

**Comment by chrisswan-rm** [comment]
No problem Daniel. Here is the video:
[https://photos.app.goo.gl/NXGTRmbXirzwQMmk8](https://photos.app.goo.gl/NXGTRmbXirzwQMmk8)

**Comment by dbuechel** [comment]
Thanks a lot for the video, @chrisswan-rm. I was now able to reproduce the issue thanks to it as I first did not click in the content area. I furthermore was able to reproduce the issue in the CEF sample application  (https://cef-builds.spotifycdn.com/index.html). Thus, the problem either lies in the implementation of the audio player itself or in the browser engine we use (CEF / Chromium). It definitively has nothing to do with SEB resp. our code as the exact same issue reproduces in the CEF sample application (one can even see the input focus jumping to the menu bar when pressing F10).

So there's most unfortunately nothing further we can do here, but I would suggest you take this up with the developers of the audio player (as to me, it does appear that the issue lies there) or you could also try to create an issue in the CEF forum / repository (but as said, I suspect the audio player and its developers should be able to quickly determine whether it indeed is an issue with their code or with the browser engine).

**Comment by chrisswan-rm** [comment]
Thanks Damian. I am almost certain that this won't be related to the audio player, as it is reproducible with two different third-party audio players and also when using AudioContext directly. Therefore this is likely an issue with CEF.

I have tried this myself in the CEF sample application, and can see the same behaviour. However, the crucial difference between the CEF sample application and SEB is that in the CEF sample app, pressing F10 **visibly** takes focus away from the page, and the "File" menu becomes selected, as in the screenshot I attached, so you could argue that stopping the audio is expected, because focus has gone away from the page. However, in SEB there is no such menu which becomes selected (maybe something is being selected behind the scenes?) Therefore I think this still is a SEB issue, as in SEB F10 should do nothing (especially if it's been blocked in the configuration file).

![image](https://user-images.githubusercontent.com/13949144/169284270-604f43e9-ed5e-4142-8e00-9be39cb24f5b.png)

**Comment by danschlet** [comment]
Browsers are usually able to play audio and video in the background. As the audio also stops in the CEF sample browser but not in Chrome, either it's really a CEF issue or it could be solved in the JS audio/video player.

There are some limitations regarding blocking of system provided key shortcuts by an application like SEB, maybe this is an example of that.

**Comment by dbuechel** [comment]
> I am almost certain that this won't be related to the audio player, as it is reproducible with two different third-party audio players and also when using AudioContext directly. Therefore this is likely an issue with CEF.

I still would recommend you to contact the developers of the audio player for the reasons already stated above. I just tested a few everyday video and audio players I know personally (in the CEF sample application), and none of them interrupt playing when pressing F10. It however may very well be just a CEF / Chromium issue, that is most certain.

> [...] you could argue that stopping the audio is expected, because focus has gone away from the page

I wouldn't quite agree with that, as e.g. listening to a podcast while working in another application is a perfectly normal use case which I reckon millions of people do. Were browsers to stop audio / video playback as soon as they loose input focus, listening to something in the background wouldn't be possible.

> [...] in SEB there is no such menu which becomes selected (maybe something is being selected behind the scenes?)

That is exactly right, yes. Due to the nature of the embedded browser engine, there is much layering of windows within windows, and the focus may be put on some element which isn't even visible. Though, it could also be that the focus would simply jump to the window menu bar if the SEB browser windows had a window menu.

> Therefore I think this still is a SEB issue, as in SEB F10 should do nothing (especially if it's been blocked in the configuration file).

It most definitely isn't, as we can reproduce the issue in the CEF sample application which has no relation to SEB whatsoever. F10 is furthermore correctly blocked, this can be tested e.g. in the live application log of SEB or any third-party application running in a SEB session. However, as @danschlet already stated above, SEB can't suppress some aspects of the operating system itself and it appears F10 is one of them, just like the ALT key keeps having its purpose and function whether SEB is running or not.

One thought that just came to my mind is that it might not have to do with the audio player directly, but more with the implementation of the playlist which triggers the audio clips consecutively? Maybe that mechanism is somehow affected by F10, thought that would be quite surprising.

**Comment by chrisswan-rm** [comment]
Just a quick update on this following some more research and testing:

- I have raised the issue with CEF, but not had a response yet: https://bitbucket.org/chromiumembedded/cef/issues/3333/audio-that-is-started-programmatically
- SEB itself is blocking F10 as expected, as we created a simple sample page that displays an alert when F10 is pressed, and this does not trigger if F10 is blocked in SEB. Therefore SEB must be preventing F10 from getting through to the web page itself, so the issue must be happening elsewhere (in the SEB application, or CEF). Therefore I don't think this can be an issue with an F10 press interacting with the audio player, as F10 is prevented from getting through to the web page at all, as expected.
- We have tried an alternative locked down browser (Trifork), and the issue does not occur in it. Interestingly, it also uses CEF, which would point to this still being a SEB issue. However, the CEF version used is different - the version of Trifork we have is using version 81.1.1.0 of libcef.dll and SEB is using 97.1.1.0, so potentially it could be an issue introduced in a newer version of CEF. I guess I could try older versions of SEB and / or CEF to prove or disprove this...

**Comment by chrisswan-rm** [comment]
Further to this, I have just built the latest version of SEB (3.3.2) against CEF 86 instead of CEF 97, and the problem is still present. Therefore I think this must be a problem with SEB after all, introduced in version 3.2.0, because:

SEB 3.1.1 (CEF 86) does not have the problem
SEB 3.2.0 (CEF 90) has the problem
SEB 3.3.2 (CEF 97) has the problem
SEB 3.3.2 **built against CEF 86, which works fine with SEB 3.1.1** has the problem

Please can you re-open this ticket and investigate? Thanks.

**Comment by dbuechel** [comment]
I'm terribly sorry, but this still doesn't prove that the issue is in SEB itself, it rather seems to indicate that the issue has been introduced around CEF 86. Can you please specify which exact versions you used (e.g. CEF 86.0.24 with Chromium 86.0.4240.198)? There are a lot of builds with major version 86, so the issue may well have been introduced in between minor or patch versions. To verify this, you could go through all the major builds of version 86 of the CEF sample application: https://cef-builds.spotifycdn.com/index.html#windows64:86 (click on "Show more builds" under "All builds").

As we already have established, the issue is reproducible with the CEF sample application, thus it is with the utmost probability an issue in the browser engine (be it CEF or Chromium itself). I hence suggest that we wait and see what the developers of CEF respond to the issue you created. I unfortunately do not have the extra time now to keep investigating this issue, especially given that we could reproduce it in CEF itself.

However, if there are new and decisive findings, I might be able to schedule this for our next sprint starting end of June. I need to mention though that this sprint already is packed full due to the planned release of SEB 3.4.0 by end of Q2 this year, so if it would turn out to be an issue in SEB after all, I cannot guarantee that we'll still be able to fix it for version 3.4.0.

**Comment by chrisswan-rm** [comment]
Thanks for the quick response Damian.

Version 3.1.1 of SEB (which does not have the issue) uses version 86.0.24.0 of CEF.

I built the source for SEB 3.3.2 against the same version of CEF (86.0.24.0) and it exhibits the issue.

**Comment by chrisswan-rm** [comment]
I managed to fix it by adding this code into LowLevelKeyboardProc in KeyboardHook.cs. This is obviously a hack, and definitely not the right way to fix it, but it shows that it can be fixed in SEB:

`if (keyData.KeyCode == 121)
				{
					SendKeys.Send("{F10}");
					return (IntPtr)1;
				}`

Basically this just repeats the F10 keystroke again. This resolves the issue, because when the issue occurs pressing F10 again (or actually any key, or just a mouse click) seems to re-activate the browser and continue Javascript execution.

**Comment by danschlet** [comment]
Not to diminish the need to fix this issue, regardless what causes it. But you mentioned "due to this issue it was found that a student can effectively pause the assessment if they press F10, which is not desirable" and @s-spann mentioned "This is quite a concern for the business as this allows manipulation of the exam".

You are aware that any computer can be put to sleep mode (for example by closing the lid of a laptop/notebook) which also pauses the exam? From my outside position it seems problematic to expect that an exam cannot be paused.

**Comment by dbuechel** [comment]
@chrisswan-rm You are right, this is neither a fix nor proof, but rather a way of programmatically performing the steps one now has to do manually to resolve the underlying issue.

Again, this still does not indicate to me that it is an issue in SEB and not in the browser engine. The `KeyboardHook` has not changed since October 1st 2019: https://github.com/SafeExamBrowser/seb-win-refactoring/commits/master/SafeExamBrowser.WindowsApi/Hooks/KeyboardHook.cs. What has changed, on the other hand, is the browser engine and its related code.

I'm terribly sorry, but as long as there isn't a clear indication (and also feedback from the CEF developers pointing in the same direction) I won't investigate this issue further due to the ongoing work I have to complete.

**Comment by chrisswan-rm** [comment]
@danschlet Yes that's a good point, we can't completely prevent a student from "pausing" an exam in some way. There are a couple of important differences though between them using this "F10" exploit to pause an exam, and putting the computer to sleep:
- An invigilator is much more likely to spot someone who has put a computer into sleep mode than someone who has used this exploit (which at a glance does not have any visual impact to what's seen on the screen).
- While putting the computer to sleep might give you more thinking time, this may not be very useful if you can no longer see the exam stimulus on the screen.

@dbuechel Yes, I agree that my fix is more of a workaround, but I think there **is** proof that the issue is with SEB and not with CEF - because SEB 3.1.1 does not have the issue, but if I build SEB 3.3.2 against exactly the same version of CEF that SEB 3.1.1 uses, the issue is still there. So while the bug itself may be in CEF, there must be something that changed in SEB itself between SEB 3.1.1 and SEB 3.2.0 to expose the bug to the user, whereas previously it did not manifest. Most likely it was some seemingly-unrelated change, rather than an obvious change to keyboard handling.

**Comment by dbuechel** [comment]
There may have been a change with side-effects, but as already stated, I do not have the time now to further investigate this issue. You're most welcome to study all the changes from version 3.1.1 to 3.2.0, they're publicly available here on GitHub. In the meantime, I suggest we wait for feedback from the CEF developers and continue from there.

**Comment by chrisswan-rm** [comment]
I did some more investigation and it looks like the change that caused this was in  this commit:
https://github.com/SafeExamBrowser/seb-win-refactoring/commit/2d1d3d9dde29144830267e0edc039564d384bf8d

Specifically, the addition of the following line to BrowserApplication.cs:
`cefSettings.CefCommandLineArgs.Add("enable-media-stream");`

This explains why the issue was still present when I built the latest SEB against an older version of CEF, because the problem seems to not be related to the CEF version, but instead happens when the "enable-media-stream" argument has been passed to CEF.

So it is likely a bug with CEF that has always been there, but it only started happening once you enabled that command line switch in SEB. I will pass this information on to CEF.

Specifically the issue seems to be that when you press F10 (or Alt) when "enable-media-stream" is enabled, the focus switches to a hidden menu and this pauses Javascript execution. If you press Enter after pressing F10 or Alt the menu itself actually appears. But this does not happen when "enable-media-stream" is disabled, which seems very bizarre.
